x448/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+pub struct SharedSecret(MontgomeryPoint);
+
+impl PublicKey {
+    /// Converts a bytes slice into a Public key
+    /// Returns None if:
+    /// -  The length of the slice is not 56
+    /// -  The point is a low order point
+    pub fn from_bytes(bytes: &[u8]) -> Option<PublicKey> {
+        let public_key = PublicKey::from_bytes_unchecked(bytes)?;
+        if public_key.0.is_low_order() {
+            return None;
+        }
+        Some(public_key)
+    }
+    /// Converts a bytes slice into a Public key
+    /// Returns None if:
+    /// -  The length of the slice is not 56
+    pub fn from_bytes_unchecked(bytes: &[u8]) -> Option<PublicKey> {
+        // First check if we have 56 bytes
+        if bytes.len() != 56 {
+            return None;
+        }
+
+        // Check if the point has low order
+        let arr = slice_to_array(bytes);
+        let point = MontgomeryPoint(arr);
+
+        Some(PublicKey(point))
+    }
+
+    /// Converts a public key into a byte slice
+    pub fn as_bytes(&self) -> &[u8; 56] {
+        self.0.as_bytes()
+    }
+}
+
+impl SharedSecret {
+    /// Converts a shared secret into a byte slice
+    pub fn as_bytes(&self) -> &[u8; 56] {
+        self.0.as_bytes()
+    }
+}
+
+impl Secret {
+    /// Generate a x448 `Secret` key.
+    // Taken from dalek-x25519
+    pub fn new<T>(csprng: &mut T) -> Self
+    where
+        T: RngCore + CryptoRng + ?Sized,
+    {
+        let mut bytes = [0u8; 56];
+
+        csprng.fill_bytes(&mut bytes);
+
+        Secret::from(bytes)
+    }
+
+    /// Clamps the secret key according to RFC7748
+    fn clamp(&mut self) {
+        self.0[0] &= 252;
+        self.0[55] |= 128;
+    }
+
+    /// Views a Secret as a Scalar
+    fn as_scalar(&self) -> MontgomeryScalar {
+        let secret = U448::from_le_slice(&self.0);
+        MontgomeryScalar::from_uint_unchecked(secret)
+    }
+
+    /// Performs a Diffie-hellman key exchange between the secret key and an external public key
+    pub fn as_diffie_hellman(&self, public_key: &PublicKey) -> Option<SharedSecret> {
+        // Check if the point is one of the low order points
+        if public_key.0.is_low_order() {
+            return None;
+        }
+        let shared_key = &public_key.0 * &self.as_scalar();
+        Some(SharedSecret(shared_key))
+    }
+
+    /// Performs a Diffie-hellman key exchange once between the secret key and an external public key
+    pub fn to_diffie_hellman(self, public_key: &PublicKey) -> Option<SharedSecret> {
+        self.as_diffie_hellman(public_key)
+    }
+
+    /// Converts a byte slice into a secret and clamp
+    pub fn from_bytes(bytes: &[u8]) -> Option<Secret> {
+        // First check if we have 56 bytes
+        if bytes.len() != 56 {
+            return None;
+        }
+
+        let secret = Secret::from(slice_to_array(bytes));
+        Some(secret)
+    }
+
+    /// Converts a secret into a byte array
+    pub fn as_bytes(&self) -> &[u8; 56] {
+        &self.0
+    }
```

Comment:
We should probably have different types for ephemeral vs static secrets ala `x25519-dalek`. For ephemeral secrets ideally the caller couldn't export this value (for forward secrecy).